packages/core/src/oidc/grants/refresh-token.ts

```diff
@@ -0,0 +1,398 @@
+ */
+export const parameters = Object.freeze(['refresh_token', 'organization_id', 'scope'] as const);
+
+/**
+ * The required parameters for the grant type.
+ *
+ * @see {@link parameters} for the full list of valid parameters.
+ */
+const requiredParameters = Object.freeze(['refresh_token'] as const) satisfies ReadonlyArray<
+  (typeof parameters)[number]
+>;
+
+// We have to disable the rules because the original implementation is written in JavaScript and
+// uses mutable variables.
+/* eslint-disable @silverhand/fp/no-let, @typescript-eslint/no-non-null-assertion, @silverhand/fp/no-mutation, unicorn/no-array-method-this-argument */
+export const buildHandler: (
+  envSet: EnvSet,
+  queries: OrganizationQueries
+  // eslint-disable-next-line complexity
+) => Parameters<Provider['registerGrantType']>['1'] = (envSet, queries) => async (ctx, next) => {
+  const { client, params, requestParamScopes, provider } = ctx.oidc;
+  const { RefreshToken, Account, AccessToken, Grant, ReplayDetection, IdToken } = provider;
+
+  assertThat(params, new InvalidGrant('parameters must be available'));
+  assertThat(client, new InvalidClient('client must be available'));
+
+  validatePresence(ctx, ...requiredParameters);
+
+  const providerInstance = instance(provider);
+  const {
+    rotateRefreshToken,
+    conformIdTokenClaims,
+    features: {
+      mTLS: { getCertificate },
+      userinfo,
+      resourceIndicators,
+    },
+  } = providerInstance.configuration();
+
+  const dPoP = await dpopValidate(ctx);
+
+  // @gao: I believe the presence of the param is validated by required parameters of this grant.
+  // Add `String` to make TS happy.
+  let refreshTokenValue = String(params.refresh_token);
+  let refreshToken = await RefreshToken.find(refreshTokenValue, { ignoreExpiration: true });
+
+  if (!refreshToken) {
+    throw new InvalidGrant('refresh token not found');
+  }
+
+  if (refreshToken.clientId !== client.clientId) {
+    throw new InvalidGrant('client mismatch');
+  }
+
+  if (refreshToken.isExpired) {
+    throw new InvalidGrant('refresh token is expired');
+  }
+
+  let cert: Optional<string>;
+  // eslint-disable-next-line @typescript-eslint/prefer-nullish-coalescing -- the original code uses `||`
+  if (client.tlsClientCertificateBoundAccessTokens || refreshToken['x5t#S256']) {
+    cert = getCertificate(ctx);
+    if (!cert) {
+      throw new InvalidGrant('mutual TLS client certificate not provided');
+    }
+  }
+
+  if (!dPoP && client.dpopBoundAccessTokens) {
+    throw new InvalidGrant('DPoP proof JWT not provided');
+  }
+
+  if (refreshToken['x5t#S256'] && refreshToken['x5t#S256'] !== certificateThumbprint(cert!)) {
+    throw new InvalidGrant('failed x5t#S256 verification');
+  }
+
+  /* === RFC 0001 === */
+  // The params object may have the key with `undefined` value, so we have to use `Boolean` to check.
+  const organizationId = cond(Boolean(params.organization_id) && String(params.organization_id));
```

Comment:
If the param may have the `undefined` value, I'm afraid `Boolean('undefined')` is a truthy value, since the param value is expected to be a string type, right?